Make some improvements to java +lsp

jsravn opened the pull request:
This is a first pass at making `+lsp` more functional:

* Fix lsp not starting automatically when opening java-mode buffers.
Placing it in the lsp-java hook does not work.
* Fix root path for lsp-jt.

jsravn commented:
I got some feedback from the lsp-java maintainer at https://github.com/emacs-lsp/lsp-java/issues/222#issuecomment-624061914. It looks like lsp-jt isn't recommended, and avy mode should be used for the code lens along w/ directly invoking dap-debug on test classes. I'll have a dig into that later.

hlissner commented on modules/lang/java/+lsp.el:
```diff
@@ -1,13 +1,18 @@
 ;;; lang/java/+lsp.el -*- lexical-binding: t; -*-
 ;;;###if (featurep! +lsp)
 
+;;;###autoload
+(defun +java/update-lsp-server ()
+  (interactive)
+  (lsp--install-server-internal (gethash 'jdtls lsp-clients) t))
```
This should be in an autoloads file, but...

hlissner commented on modules/lang/java/+lsp.el:
```diff
@@ -1,13 +1,18 @@
+  (lsp--install-server-internal (gethash 'jdtls lsp-clients) t))
+
+(add-hook! java-mode-local-vars #'lsp!)
+(map! :map java-mode-map
+      :localleader
+      :desc "Update the Eclipse LSP server" "U" #'+java/update-lsp-server)
```
Please remove the `map!`. I don't think a update-lsp-server keybind is necessary. At least, it shouldn't be so hardcoded to a single client. A more general `+lsp/update-running-server` command in the `:tools lsp` module would be better.

What's more, I don't think it should be run so often that it warrants a keybind.

jsravn replied:
Makes sense.

hlissner commented on modules/lang/java/+lsp.el:
```diff
@@ -1,13 +1,18 @@
+
 (use-package! lsp-java
-  :after lsp-clients
-  :hook (java-mode-local-vars . lsp!)
```
The reason this did not work was due to how `use-package` processes hooks when an `:after` is involved. Please move the `add-hook` call into the use-package! block's `:preface`. I'd like to keep lsp-java's config inside it.

```elisp
(use-package! lsp-java
  :after lsp-clients
  :preface
  (setq lsp-java-server-install-dir (concat doom-etc-dir "eclipse.jdt.ls/server/")
        lsp-java-workspace-dir (concat doom-etc-dir "java-workspace"))

  (add-hook 'java-mode-local-vars-hook #'lsp!))
```

jsravn commented:
Addressed the review comments, and also fix `lsp-jt-root` so java-test stuff works (in case someone wants to experiment with it, as it is in a mostly working state).

hlissner commented:
Thanks again for your help!
